### DISCUSSION TITLE
How to correct formula: Set width of selected nodes

### DISCUSSION THREAD

**Discussion by Illegit** [freeplane/freeplane#267] [Help] [answered]
I ask for a suggestion how to correct the following script that has to set the width of a selection of nodes, it generates a recurrence that I like to avoid: it is asking the same questions for all of the selected nodes, instead I need just one time question and after sizing the nodes (which command is wrong?)

import org.freeplane.view.swing.map.NodeView
//import org.freeplane.features.map.NodeModel
//import org.freeplane.features.mode.Controller
	customWidth = ui.showInputDialog(node.delegate , 'Set node width (pixel)', '')
def nods = c.selecteds
	nods*.style*.minNodeWidth = customWidth as int
    nods*.style*.maxNodeWidth = customWidth as int

Thanks
SFPC

**Comment by Illegit** [comment]
Comment: In fact in the dialog input command I do not know even the meaning of the NodeModel concept...

**Comment by macmarrum** [comment]
Hi Illegit/SFPC

A great question. This is something that many scripters run into.

The topic is related to how a script is executed by Freeplane, e.g. on a single node or on all selected nodes. Please see https://www.freeplane.org/wiki/index.php/Scripting#Execution_modes

In your case, you want Freeplane to run the script only once, because your code takes care of going over all selected nodes and performing the width adjustments. Putting `// @ExecutionModes({ON_SINGLE_NODE})` in your script (usually at the beginning) will inform Freeplane to run it only once, regardless of how many nodes are selected.

**Reply by Illegit** [reply]
Hi macmarrum, thankyou, done, but still if I select 4 nodes it asks the same question 4 times, why? I need just only once the question.
SFPC

**Comment by macmarrum** [comment, marked as answer]
For me the input dialog shows only once.

I saved your code in a file `<freeplane-user-dir>/1.9.x/scripts/SFPC.groovy`, adding `// @ExecutionModes({ON_SINGLE_NODE})` on top of the file → [SFPC.groovy.zip](https://github.com/freeplane/freeplane/files/8309655/SFPC.groovy.zip)
I restarted my Freeplane.
I selected 4 nodes and ran the script (`Tools->Scripts->SFPC`)

How do you run it?

**Comment by Illegit** [comment]
Hi macmarrum, I forgot to restart FP. it seems FP kept in memory previous settings. This is a little bit tricky because some changes in the script were accepted and this one was not... After restart everything si as expected.
Thank you again.
SFPC

**Comment by EdoFro** [comment]
Hi,
Just an additional information:

Sometimes, for example when your script moves or reorders the selected nodes, 
the script *changes* which are the selected nodes while running the code and gives unwanted results or throws errors.

For those cases you can create an empty list and *add* the selected nodes to it. Then the script will remeber which were the selected nodes at the beginning of the script and it won`t confuse itself.

in short:
change this line:
`def nods = c.selecteds`

to this:
`def nods = [ ] + c.selecteds`

----
**Edit:**

`def nods = [ ] << c.selecteds` 
adds the nodes as a list inside the list. That produces problems. (see next posts)

use this instead:
`def nods = [ ] + c.selecteds`    adds the nodes to the list.




bye,

edo

**Reply by Illegit** [reply]
I got this impression, and I was thinking to be wrong (maybe messing something): very well.
Thanks
SFPC

**Comment by Illegit** [comment]
But Edo I got an "error type mismatch"...?
SFPC

**Reply by EdoFro** [reply]
sorry, my fault 
when using `<<` it adds all the nodes as one item in the list (it ends as a list inside a list)
use instead:
 
`def nods = [ ]  + c.selecteds`

**Reply by Illegit** [reply]
Ok, now is working. Thankyou.
SFPC
